### PR TITLE
feat: contract and e2e tests

### PR DESCRIPTION
**Pull request by TheAustrian1998**
- Functional contracts
- e2e tests (I'm not sure if integration and unit test are needed because everything is tested in e2e, let me know)

If all good I'll proceed with deploy script, README, and docs.

### CODE REVIEW

File: solidity/ExpirableAirdrop.sol [hunk 1]
```diff
@@ -1,60 +1,106 @@
...
   /// @notice Creates a new ExpirableAirdrop contract
   /// @param _token token to airdrop
+  /// @param _expirationTimestamp expiration timestamp
   /// @param _merkleRoot of claimees
   constructor(
     IERC20 _token,
     uint256 _expirationTimestamp,
     bytes32 _merkleRoot
-  ) {
+  ) Governable(msg.sender) {
```
Comment (nchamo) [inline solidity/ExpirableAirdrop.sol]: Could we make the governable configurable and take it in the constructor?

File: solidity/ExpirableAirdrop.sol [hunk 2]
```diff
@@ -1,60 +1,106 @@
...
+  event Retrieved(uint256 timestamp, address receiver);
 
   /// @notice Creates a new ExpirableAirdrop contract
   /// @param _token token to airdrop
+  /// @param _expirationTimestamp expiration timestamp
   /// @param _merkleRoot of claimees
   constructor(
     IERC20 _token,
     uint256 _expirationTimestamp,
     bytes32 _merkleRoot
```
Comment (nchamo) [inline solidity/ExpirableAirdrop.sol:38]: Let's do this in a different pull request, but I think this should be configurable (updatable). Based on [0xngmi's reflections](https://github.com/tubby-cats/tubby-launch-reflections), we might need to update it if we make a mistake when building the tree

**Review by nchamo — COMMENTED** [review]
I haven't looked at the tests yet, but left some early feedback 😄 

Looking great so far!

File: solidity/ExpirableAirdrop.sol (hunk 1 above)
Comment (TheAustrian1998) [reply solidity/ExpirableAirdrop.sol]: Fixed!

File: solidity/ExpirableAirdrop.sol [hunk 3]
```diff
@@ -1,60 +1,106 @@
...
 
     // EFFECTS
-    hasClaimed[to] = true;
+    hasClaimed[claimee] = true;
+
+    // INTERACTIONS
+    token.transfer(receiver, amount);
+
+    emit Claimed(claimee, receiver, amount);
+  }
```
[the remark this answers is not in the document]
Comment (nchamo) [reply solidity/ExpirableAirdrop.sol]: Agree with this, let's use `safeTransferFrom` from OZ, similar to this
https://github.com/Mean-Finance/transformers/blob/main/solidity/contracts/transformers/ERC4626Transformer.sol#L12

[the remark this answers is not in the document]
Comment (nchamo) [reply solidity/ExpirableAirdrop.sol]: Well, that's annoying... Maybe we should emit before transfering?

[the remark this answers is not in the document]
Comment (nchamo) [reply solidity/ExpirableAirdrop.sol]: You can ignoring this by adding
`//slither-disable-next-line timestamp` before the function is created

File: solidity/ExpirableAirdrop.sol [hunk 4]
```diff
@@ -1,60 +1,106 @@
...
+
+  /// @notice Deposit tokens to airdrop
+  /// @param amount of tokens to deposit
+  function depositTokens(uint256 amount) external {
+    // INTERACTIONS
+    token.transferFrom(msg.sender, address(this), amount);
+
+    emit Deposited(amount);
+  }
```
[the remark this answers is not in the document]
Comment (nchamo) [reply solidity/ExpirableAirdrop.sol]: Same with `safeTransferFrom`

File: solidity/ExpirableAirdrop.sol [hunk 5]
```diff
@@ -1,60 +1,108 @@
...
+    // CHECKS
+    if (block.timestamp <= expirationTimestamp) revert NotExpired();
 
     // INTERACTIONS
-    token.transfer(to, amount);
+    address governor = IGovernable(address(this)).governor();
+    token.transfer(governor, token.balanceOf(address(this)));
 
-    emit Claimed(to, amount);
+    emit Retrieved(block.timestamp, governor);
```
Comment (nchamo) [inline solidity/ExpirableAirdrop.sol]: We don't need to emit the block's timestamp, that is already calculable by looking at the transaction, and it costs gas to emit it

File: solidity/ExpirableAirdrop.sol [hunk 6]
```diff
@@ -1,60 +1,108 @@
...
   }
 
-  /// @notice Allows claiming tokens if address is part of merkle tree
-  /// @param to address of claimee
-  /// @param amount of tokens owed to claimee
-  /// @param proof merkle proof to prove address and amount are in tree
-  function claim(
-    address to,
+  function _claim(
+    address claimee,
```
Comment (nchamo) [inline solidity/ExpirableAirdrop.sol]: We should try to be consistent with other parts of the code. Parameters and interval variables should start with `_`

File: solidity/ExpirableAirdrop.sol [hunk 7]
```diff
@@ -1,60 +1,108 @@
...
+
+    emit Claimed(claimee, receiver, amount);
+  }
+
+  /// @notice Allows claiming tokens, if address is part of merkle tree
+  /// @param claimee address of claimee
+  /// @param amount of tokens owed to claimee
+  /// @param proof merkle proof to prove address and amount are in tree
+  function claim(
+    address claimee,
```
Comment (nchamo) [inline solidity/ExpirableAirdrop.sol]: Same with starting with `_`

File: solidity/ExpirableAirdrop.sol [hunk 8]
```diff
@@ -1,60 +1,108 @@
...
+  ) external {
+    _claim(claimee, claimee, amount, proof);
+  }
+
+  /// @notice Allows claiming tokens and send tokens to a receiver, if address is part of merkle tree
+  /// @param receiver address to send tokens
+  /// @param amount of tokens owed to claimee
+  /// @param proof merkle proof to prove address and amount are in tree
+  function claimAndTransfer(
+    address receiver,
```
Comment (nchamo) [inline solidity/ExpirableAirdrop.sol]: Same with starting with `_`

File: solidity/ExpirableAirdrop.sol [hunk 9]
```diff
@@ -1,60 +1,108 @@
...
+    uint256 amount,
+    bytes32[] calldata proof
+  ) external {
+    _claim(msg.sender, receiver, amount, proof);
+  }
+
+  /// @notice Deposit tokens to airdrop
+  /// @param amount of tokens to deposit
+  function depositTokens(uint256 amount) external {
```
Comment (nchamo) [inline solidity/ExpirableAirdrop.sol]: Is this really necessary? We can transfer the tokens to the contract, I think that approving + calling this function would just be more expensive. What do you think @TheAustrian1998 ?

File: solidity/ExpirableAirdrop.sol [hunk 10]
```diff
@@ -1,60 +1,108 @@
...
+    // INTERACTIONS
+    token.transferFrom(msg.sender, address(this), amount);
+
+    emit Deposited(amount);
+  }
+
+  /// @notice Return unclaimed tokens to governor
+  function retrieveTokens() external {
```
Comment (nchamo) [inline solidity/ExpirableAirdrop.sol]: Just a suggestion, feel free to ignore
```suggestion
  function retrieveUnclaimedTokens() external {
```

**Review by nchamo — COMMENTED** [review]
Adding some more comments

File: solidity/ExpirableAirdrop.sol (hunk 9 above)
Comment (TheAustrian1998) [reply solidity/ExpirableAirdrop.sol]: I made it that way to prevent sending tokens to another contract or EOA by mistake, but I can remove it.

File: solidity/ExpirableAirdrop.sol (hunk 10 above)
Comment (TheAustrian1998) [reply solidity/ExpirableAirdrop.sol]: You are right! That name will clear some confusions about function behavior.

File: codechecks.yml [hunk 11]
```diff
@@ -1,6 +1,3 @@
-checks:
```
Comment (0xged) [inline codechecks.yml:1]: If there is not going to be a gas-report workflow, you can remove this file 😊  - its the configuration for codechecks that its only used in the github action `gas-report`.

File: package.json [hunk 12]
```diff
@@ -58,7 +59,8 @@
     "eth-gas-reporter/colors": "1.4.0"
   },
   "dependencies": {
-    "@openzeppelin/contracts": "4.7.1"
+    "@openzeppelin/contracts": "4.7.1",
+    "merkletreejs": "^0.2.32"
```
Comment (0xged) [inline package.json]:
```suggestion
    "merkletreejs": "0.2.32"
```
We usually set "exact" version on dependencies, just to be sure that everyone is working on the same version of packages (although `yarn.lock` should take care of that too).

File: solidity/utils/Governable.sol [hunk 13]
```diff
@@ -3,7 +3,7 @@
 pragma solidity >=0.7.0;
```
Comment (0xged) [inline solidity/utils/Governable.sol]: I'm an a-hole and pushed an old version of Governable. You can find the new one here: https://github.com/Mean-Finance/transformers/blob/main/solidity/contracts/utils/Governable.sol

Sorry for the extra work 🤕

File: solidity/ExpirableAirdrop.sol [hunk 14]
```diff
@@ -1,60 +1,111 @@
...
   bytes32 public immutable merkleRoot;
   mapping(address => bool) public hasClaimed;
 
   error AlreadyClaimed();
   error Expired();
+  error NotExpired();
   error NotInMerkle();
 
-  event Claimed(address indexed to, uint256 amount);
+  event Claimed(address indexed _claimee, address indexed _receiver, uint256 _amount);
```
Comment (0xged) [inline solidity/ExpirableAirdrop.sol]:
```suggestion
  event Claimed(address _claimee, address _receiver, uint256 _amount);
```
I'd avoid indexing event params, since we are going to use `the graph` and emitting indexed stuff is more costly (from what I remember).

File: solidity/ExpirableAirdrop.sol [hunk 15]
```diff
@@ -1,60 +1,111 @@
...
-    hasClaimed[to] = true;
+    hasClaimed[_claimee] = true;
+
+    // INTERACTIONS
+    token.safeTransfer(_receiver, _amount);
+
+    emit Claimed(_claimee, _receiver, _amount);
+  }
+
+  /// @notice Allows claiming tokens, if address is part of merkle tree
```
Comment (0xged) [inline solidity/ExpirableAirdrop.sol]:
```suggestion
  /// @notice Allows claiming tokens for someone else but keeping the owner of the airdrop as the receiver. Will only work if address is part of merkle tree
```
Just so its extra clear 😋

File: solidity/ExpirableAirdrop.sol [hunk 16]
```diff
@@ -1,60 +1,111 @@
...
+    address _receiver,
+    uint256 _amount,
+    bytes32[] calldata _proof
+  ) external {
+    _claim(msg.sender, _receiver, _amount, _proof);
+  }
+
+  /// @notice Deposit tokens to airdrop
+  /// @param _amount of tokens to deposit
+  function depositTokens(uint256 _amount) external {
```
Comment (0xged) [inline solidity/ExpirableAirdrop.sol:93]: I was going to argue this was not entirely necessary, but then I saw you guys comments. Seems sensible to have the `depositTokens` function, specially on Optimism that everything is pretty cheap.

**Review by 0xged — COMMENTED** [review]
Amazing work so far! Some misc suggestions to be honest. I'll be reviewing the tests later today :)

File: solidity/ExpirableAirdrop.sol (hunk 16 above)
Comment (nchamo) [reply solidity/ExpirableAirdrop.sol:93]: Yeah, agree. Makes sense to keep it, and we can decide if we want to use it or not later

Comment (TheAustrian1998) [reply solidity/ExpirableAirdrop.sol:93]: Cool, I'll keep it

File: solidity/ExpirableAirdrop.sol [hunk 17]
```diff
@@ -1,60 +1,110 @@
...
+    token.safeTransfer(_receiver, _amount);
+
+    emit Claimed(_claimee, _receiver, _amount);
+  }
+
+  /// @notice Allows claiming tokens for someone else but keeping the owner of the airdrop as the receiver. Will only work if address is part of merkle tree
+  /// @param _claimee address of claimee
+  /// @param _amount of tokens owed to claimee
+  /// @param _proof merkle proof to prove address and amount are in tree
+  function claim(
```
Comment (nchamo) [inline solidity/ExpirableAirdrop.sol]: Like suggested, we can rename this to `claimAndSendToClaimee`

File: test/e2e/expirable-airdrop.spec.ts [hunk 18]
```diff
@@ -0,0 +1,188 @@
...
+  before(async () => {
+    nowTimestamp = await time.latest();
+    expirationTimestamp = nowTimestamp + 3 * oneMonth; // 3 months
+    [deployer, ...accounts] = await ethers.getSigners();
+
+    [root, tree] = createMerkleRoot(accounts, airdropAmount);
+
+    // deploy erc20 mock
+    const Token = await ethers.getContractFactory('ERC20Mock');
+    token = await Token.deploy('Token', 'TKN');
```
Comment (nchamo) [inline test/e2e/expirable-airdrop.spec.ts]:
```suggestion
    const tokenFactory = await ethers.getContractFactory('ERC20Mock');
    tokenFactory = await Token.deploy('Token', 'TKN');
```

File: test/e2e/expirable-airdrop.spec.ts [hunk 19]
```diff
@@ -0,0 +1,188 @@
...
+
+    // deploy contract
+    const ExpirableAirdrop = await ethers.getContractFactory('ExpirableAirdrop');
+    expirableAirdrop = await ExpirableAirdrop.deploy(deployer.address, token.address, expirationTimestamp, root);
```
Comment (nchamo) [inline test/e2e/expirable-airdrop.spec.ts]:
```suggestion
    const expirableAirdropFactory = await ethers.getContractFactory('ExpirableAirdrop');
    expirableAirdrop = await expirableAirdropFactory.deploy(deployer.address, token.address, expirationTimestamp, root);
```

File: test/e2e/expirable-airdrop.spec.ts [hunk 20]
```diff
@@ -0,0 +1,188 @@
...
+
+    then('airdrop is deposited', async () => {
+      // approve
+      await token.approve(expirableAirdrop.address, ethers.constants.MaxUint256);
+
+      // deposit
+      await expirableAirdrop.depositTokens(toDeposit);
+
+      // save snapshot for later
+      snapshot = await takeSnapshot();
```
Comment (nchamo) [inline test/e2e/expirable-airdrop.spec.ts]: It's very hard for me to understand what the tests are doing if we are linking them via the snapshot. If we were to change the order, then they would break, right?

What we normally do is:
- We take a snapshot at the end of the `before`
- We restore it in a `beforeEach`

By doing so, the context is always the same for all tests, they can be re-ordered, and they need to explicitely configure the context that they want to test. It might be a little more verbose, but I personally prefer clarity over shortness in tests

File: test/e2e/expirable-airdrop.spec.ts [hunk 21]
```diff
@@ -0,0 +1,188 @@
...
+  let nowTimestamp: number;
+  const oneMonth: number = 30 * 24 * 60 * 60; // one month in seconds
+  let expirationTimestamp: number;
+  let accounts: any[];
+  let deployer: any;
+  let root: any;
+  let tree: any;
+  let token: any;
+  let expirableAirdrop: any;
+  let airdropAmount: number = 10;
```
Comment (nchamo) [inline test/e2e/expirable-airdrop.spec.ts:38]: I see that you are using `toUnit(airdropAmount)` in a lot of places. Maybe we could have another variable here that is `toUnit(airdropAmount)` and use that one?

File: test/e2e/expirable-airdrop.spec.ts [hunk 22]
```diff
@@ -0,0 +1,188 @@
...
+      // save snapshot for later
+      snapshot = await takeSnapshot();
+
+      // expect
+      expect(await token.balanceOf(expirableAirdrop.address)).equal(toDeposit);
+    });
+  });
+
+  when('claim', () => {
+    let alice: any;
```
Comment (nchamo) [inline test/e2e/expirable-airdrop.spec.ts]: Could we add some types? For example, alice could be `SignerWithAddress`

File: test/e2e/expirable-airdrop.spec.ts [hunk 23]
```diff
@@ -0,0 +1,188 @@
...
+      expect(await token.balanceOf(alice.address)).equal(toUnit(airdropAmount));
+    });
+
+    then('revert if already claimed', async () => {
+      await expect(expirableAirdrop.connect(alice).claim(alice.address, toUnit(airdropAmount), proof)).to.be.revertedWithCustomError(
+        expirableAirdrop,
+        'AlreadyClaimed'
+      );
+    });
+
```
Comment (nchamo) [inline test/e2e/expirable-airdrop.spec.ts:174]: 1. Could we add a test where another address claims in behalf if Alice, and the tokens are sent to her?
2. Could we add a test that reverts when the claim has expired?

File: test/e2e/expirable-airdrop.spec.ts [hunk 24]
```diff
@@ -0,0 +1,188 @@
...
+    then('airdrop claimed', async () => {
+      // user claims airdrop
+      await expirableAirdrop.connect(bob).claimAndTransfer(randomAddress2, toUnit(airdropAmount), proof);
+
+      // expect
+      expect(await token.balanceOf(randomAddress2)).equal(toUnit(airdropAmount));
+    });
+
```
Comment (nchamo) [inline test/e2e/expirable-airdrop.spec.ts:233]: 1. Could we add a test that reverts if already claimed?
2. Could we add a test that reverts if the user doesn't have something to claim?

I know that technically that was tested on `claim`, but I think it takes only a little time, and I prefer if the test doesn't know that they are implemented with the same internal function

**Review by nchamo — COMMENTED** [review]
Added some more comments to the tests 😊

File: test/e2e/expirable-airdrop.spec.ts [hunk 25]
```diff
@@ -0,0 +1,250 @@
...
+    // save snapshot for later
+    snapshot = await takeSnapshot();
+  });
+
+  beforeEach(async () => {
+    await snapshot.restore();
+  });
+
+  when('deposit', () => {
```
Comment (nchamo) [inline test/e2e/expirable-airdrop.spec.ts]: Could we test that the `Deposited` event is emmited?

I would suggest something like this:
```
when(’deposit is executed, () => {
    const toDeposit = airdropAmountBN.mul(accounts.length + 1)
    let tx: TransactionResponse
    given(() => {
        await token.mint(deployer.address, toDeposit);
        await token.approve(expirableAirdrop.address, ethers.constants.MaxUint256);
        tx = await expirableAirdrop.depositTokens(toDeposit);
    })
    then('tokens are deposited', () => {
        // Check that balance changed
    })
    then(‘event is emitted’, () => {
        expect(tx).to.have.emitted(…)
    }) 
})
```

File: test/e2e/expirable-airdrop.spec.ts [hunk 26]
```diff
@@ -0,0 +1,250 @@
...
+
+      // deposit
+      await expirableAirdrop.depositTokens(toDeposit);
+
+      // expect
+      expect(await token.balanceOf(expirableAirdrop.address)).equal(toDeposit);
+    });
+  });
+
+  when('claim and send to claimee', () => {
```
Comment (nchamo) [inline test/e2e/expirable-airdrop.spec.ts]: I would suggest a slightly different approach. Something like this:
```
describe('claimAndSendToClaimee', () => {
    let leaf: any;
    let proof: any;
    given(async () => {
        await prepareAirdrop();
        leaf = getLeaf(alice.address, airdropAmountBN);
        proof = tree.getHexProof(leaf);
    });
    when('airdrop is claimed by claimee', () => {
        let tx: TransactionResponse
        given(() => {
            tx = await expirableAirdrop.connect(alice).claimAndSendToClaimee(alice.address, airdropAmountBN, proof);
        })
        then('tokens are transferred to claimee', () => {
            // Checks balance
        })
        then('event is emitted', () => {
            // Checks event in tx
        })
        then('claimee is marked as claimed', () => {
            // Checks that claimee is marked as claimed
        })
    }
    when('airdrop is claimed on behalf of claimee', () => {
        ...
    })
    when('airdrop is already claimed', () => {
        given(() => {
            // Claim once
        })
        then('claiming again will result in a revert', () => {
            ...
        })
    })
})
```
I believe this will make tests clearer and easier to read, since there is a clear separation of what is executed vs what the result is expected to be

In some cases it can lead to repeated code. For example, in this case, the assertions in the first two tests would be the same. In those cases, you can create a function like this:
```
function thenAirdropIsClaimed(claimee: string, reciever: string, amount: BigNumberish) {
    then('tokens are transferred to claimee', () => {
        // Checks balance
    })
    then('event is emitted', () => {
        // Checks event in tx
    })
    then('claimee is marked as claimed', () => {
        // Checks that claimee is marked as claimed
    })
}
```
And you can simply call it from inside those two tests. 

I know it can be a lot, and I know that I'm very special about tests, but I've learned that it's very important that they are clear and readable. By doing so, it should take 5 seconds to map assertions to state modifications, and you will be able to tell if you are testing everything you should be testing

**Review by nchamo — COMMENTED** [review]
Please don't hate me 😅

**Review by nchamo — APPROVED** [review]
AMAAAAAAZING WORK!

I love the end result @TheAustrian1998 , thank you for taking the time to do this ❤️

File: test/e2e/expirable-airdrop.spec.ts [hunk 27]
```diff
@@ -0,0 +1,302 @@
...
+    token = await TokenFactory.deploy('Token', 'TKN');
+
+    // deploy contract
+    const ExpirableAirdropFactory = await ethers.getContractFactory('ExpirableAirdrop');
+    expirableAirdrop = await ExpirableAirdropFactory.deploy(deployer.address, token.address, expirationTimestamp, root);
+
+    // load accounts
+    alice = accounts[1];
+    bob = accounts[2];
+    carl = accounts[3];
```
Comment (0xged) [inline test/e2e/expirable-airdrop.spec.ts]:
```suggestion
    [,alice, bob, carl] = accounts;
```
I think this should work also 😋

File: test/e2e/expirable-airdrop.spec.ts [hunk 28]
```diff
@@ -0,0 +1,302 @@
...
+
+    // save snapshot for later
+    snapshot = await takeSnapshot();
+  });
+
+  beforeEach(async () => {
+    await snapshot.restore();
+  });
+
+  when('depositTokens()', () => {
```
Comment (0xged) [inline test/e2e/expirable-airdrop.spec.ts]:
```suggestion
  describe('depositTokens()', () => {
```
Describe is used to describe how a function will work, while `when` is used to describe context. For example:
```
describe 'Drink a coke'
    when 'bottle is empty'
       then 'nothing comes out'
```

File: test/e2e/expirable-airdrop.spec.ts [hunk 29]
```diff
@@ -0,0 +1,302 @@
...
+      expect(await token.balanceOf(expirableAirdrop.address)).equal(toDeposit);
+    });
+
+    then('event is emitted', async () => {
+      // expect
+      expect(tx).to.have.emit(expirableAirdrop, 'Deposited').withArgs(toDeposit);
+    });
+  });
+
+  when('claimAndSendToClaimee()', () => {
```
Comment (0xged) [inline test/e2e/expirable-airdrop.spec.ts]: Same as above 😊

File: test/e2e/expirable-airdrop.spec.ts [hunk 30]
```diff
@@ -0,0 +1,302 @@
...
+
+    when('airdrop is claimed by claimee', () => {
+      let tx: Transaction;
+
+      given(async () => {
+        tx = await expirableAirdrop.connect(alice).claimAndSendToClaimee(alice.address, airdropAmountBN, proof);
+      });
+
+      then('tokens are transferred to claimee', async () => {
+        // expect
```
Comment (0xged) [inline test/e2e/expirable-airdrop.spec.ts]:
```suggestion

```
This will apply to all //expect comment, unless there is something I am missing 😋

File: test/e2e/expirable-airdrop.spec.ts [hunk 31]
```diff
@@ -0,0 +1,302 @@
...
+
+    then('revert if no airdrop', async () => {
+      leaf = getLeaf(randomAddress, airdropAmountBN);
+      proof = tree.getHexProof(leaf);
+
+      await expect(expirableAirdrop.connect(random).claimAndSendToClaimee(randomAddress, airdropAmountBN, proof)).to.be.revertedWithCustomError(
+        expirableAirdrop,
+        'NotInMerkle'
+      );
+    });
```
Comment (0xged) [inline test/e2e/expirable-airdrop.spec.ts]:
```suggestion
   when('there is no airdrop', () => {
      then('tx gets reverted with custom error', async () => {
        leaf = getLeaf(randomAddress, airdropAmountBN);
        proof = tree.getHexProof(leaf);

        await expect(expirableAirdrop.connect(random).claimAndSendToClaimee(randomAddress, airdropAmountBN, proof)).to.be.revertedWithCustomError(
          expirableAirdrop,
          'NotInMerkle'
        );
      });
   });
```

I'd rather have this naming and `describe, given, when then`  🚀 . Applies for the same test below.

File: test/e2e/expirable-airdrop.spec.ts [hunk 32]
```diff
@@ -0,0 +1,302 @@
...
+      await prepareAirdrop();
+      balanceBefore = await token.balanceOf(expirableAirdrop.address);
+    });
+
+    then('revert if not expired yet', async () => {
+      // try to retrieve
+      await expect(expirableAirdrop.retrieveUnclaimedTokens()).to.be.revertedWithCustomError(expirableAirdrop, 'NotExpired');
+    });
+
+    when('tokens back in governor wallet', () => {
```
Comment (0xged) [inline test/e2e/expirable-airdrop.spec.ts]:
```suggestion
    when('airdrop has not expired', () => {
    
       then('tx gets reverted with custom error', async () => {
         // try to retrieve
         await expect(expirableAirdrop.retrieveUnclaimedTokens()).to.be.revertedWithCustomError(expirableAirdrop, 'NotExpired');
       });
    });

    when('airdrop has expired', () => {
```

File: test/e2e/expirable-airdrop.spec.ts [hunk 33]
```diff
@@ -0,0 +1,302 @@
...
+
+      then('contract balance before retrieving is more than zero', async () => {
+        // expect
+        expect(Number(balanceBefore)).greaterThan(0);
+      });
+
+      then('tokens are transferred to governance', async () => {
+        // expect
+        expect(await token.balanceOf(deployer.address)).equal(balanceBefore);
+      });
```
Comment (0xged) [inline test/e2e/expirable-airdrop.spec.ts]: There is no need of testing the contract balance before, but you might use the `balanceBefore` to compare with the `balanceAfter`.
For example:
- ContractsBalanceNow = contractBalanceBefore + deployerBalanceNow

Kinda testing the sanity between balances of the contract and the one receiving the unclaimed airdrop.

**Review by 0xged — APPROVED** [review]
I've suggested a few **naming** changes and order on the tests, but its mostly miscellaneous. 
Amazing improvement on each commit, this tests look fine and dandy. Great ffff* work @TheAustrian1998! 🥇
